// @douyinfe/semi-ui/lib/es/aiChatInput/extension/inputSlot/component.js
import { NodeViewWrapper, NodeViewContent } from '@tiptap/react';
import React, { useRef, useLayoutEffect, useState } from 'react';
import { strings } from '@douyinfe/semi-foundation/lib/es/aiChatInput/constants';
const InputSlotComponent = props => {
  const isEmpty = props.node.textContent === strings.ZERO_WIDTH_CHAR;
  const placeholder = props.node.attrs.placeholder || '';
  const placeholderRef = useRef(null);
  const [placeholderWidth, setPlaceholderWidth] = useState(undefined);
  useLayoutEffect(() => {
    if (isEmpty && placeholderRef.current) {
      const timer = setTimeout(() => {
        var _a;
        setPlaceholderWidth((_a = placeholderRef.current) === null || _a === void 0 ? void 0 : _a.offsetWidth);
      });
      return () => {
        clearTimeout(timer);
      };
    }
    return undefined;
  }, [isEmpty, placeholder]);
  return /*#__PURE__*/React.createElement(NodeViewWrapper, {
    as: "span",
    className: "input-slot",
    style: {
      minWidth: isEmpty && placeholderWidth ? `${placeholderWidth}px` : undefined
    }
  }, /*#__PURE__*/React.createElement("span", {
    ref: placeholderRef,
    "data-placeholder": true,
    contentEditable: false,
    className: "input-slot-placeholder",
    style: {
      display: isEmpty ? undefined : 'none'
    }
  }, placeholder), /*#__PURE__*/React.createElement(NodeViewContent, {
    as: "span",
    className: "content"
  }));
};
export default InputSlotComponent;
